# Proyek Analisis Data: E-Commerce Public Dataset
- **Nama:** Ridwan Setio Budi
- **Email:** ridwansetiobudi77@gmail.com
- **ID Dicoding:** ridwanstbd

## Menentukan Pertanyaan Bisnis

- Pertanyaan 1
- Pertanyaan 2

## Import Semua Packages/Library yang Digunakan

In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Data Wrangling

### Gathering Data

In [23]:
sellers_df = pd.read_csv("./data/sellers_dataset.csv")
sellers_df.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [24]:
product_categories = pd.read_csv("./data/product_category_name_translation.csv")
product_categories.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [25]:
products_df = pd.read_csv("./data/products_dataset.csv")
products_df.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [26]:
customers_df = pd.read_csv("./data/customers_dataset.csv")
customers_df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [27]:
geolocations_df = pd.read_csv("./data/geolocation_dataset.csv")
geolocations_df.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [28]:
orders_df = pd.read_csv("./data/orders_dataset.csv")
orders_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [29]:
order_items_df = pd.read_csv("./data/order_items_dataset.csv")
order_items_df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [30]:
order_payments_df = pd.read_csv("./data/order_payments_dataset.csv")
order_payments_df.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [31]:
order_reviews_df = pd.read_csv("./data/order_reviews_dataset.csv")
order_reviews_df.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

### Assessing Data

In [36]:
sellers_df.describe()
# Melihat frekuensi kode pos untuk mendeteksi anomali
sellers_df['seller_zip_code_prefix'].value_counts().head(10)

seller_zip_code_prefix
14940    49
13660    10
16200     9
13920     9
14020     8
1026      8
87050     8
13481     7
37540     7
18015     6
Name: count, dtype: int64

In [41]:
# Periksa penjual dengan kode pos 14940
sellers_14940 = sellers_df[sellers_df['seller_zip_code_prefix'] == 14940]
print(f"Jumlah penjual unik dengan kode pos 14940: {sellers_14940['seller_id'].nunique()}")

# Periksa format kode pos
digits_count = sellers_df['seller_zip_code_prefix'].astype(str).str.len().value_counts()
print("\nDistribusi panjang digit kode pos:")
print(digits_count)

Jumlah penjual unik dengan kode pos 14940: 49

Distribusi panjang digit kode pos:
seller_zip_code_prefix
5    2068
4    1027
Name: count, dtype: int64


In [42]:
# Cek apakah kode pos 4 digit terdapat dalam dataset geolokasi
four_digit_zips = sellers_df[sellers_df['seller_zip_code_prefix'].astype(str).str.len() == 4]['seller_zip_code_prefix'].unique()
matching_geo = geolocations_df[geolocations_df['geolocation_zip_code_prefix'].isin(four_digit_zips)]
print(f"Jumlah kode pos 4 digit yang ditemukan dalam dataset geolokasi: {matching_geo['geolocation_zip_code_prefix'].nunique()}")

Jumlah kode pos 4 digit yang ditemukan dalam dataset geolokasi: 829


In [43]:
print("Detail lokasi untuk kode pos 14940:")
print(sellers_14940[['seller_city', 'seller_state']].drop_duplicates())

Detail lokasi untuk kode pos 14940:
   seller_city seller_state
64    ibitinga           SP


**Insight:**
- Semua 49 penjual dengan kode pos 14940 berlokasi di kota Ibitinga, negara bagian São Paulo (SP), Brasil.
- Fakta bahwa 49 penjual (sekitar 1,58% dari total 3095 penjual) terkonsentrasi di satu lokasi kecil adalah fenomena bisnis yang menarik, bukan anomali data.

In [38]:
geolocations_df.info()
geolocations_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng
count                 1.000163e+06     1.000163e+06     1.000163e+06
mean                  3.657417e+04    -2.117615e+01    -4.639054e+01
std                   3.054934e+04     5.715866e+00     4.269748e+00
min                   1.001000e+03    -3.660537e+01    -1.014668e+02
25%                   1.107500e+04    -2.360355e+01    -4.857317e+01
50%                   2.653000e+04    -2.291938e+01    -4.663788e+01
75%                   6.350400e+04    -1.997962e+01    -4.376771e+01
max                   9.999000e+04     4.506593e+01     1.211054e+02

In [39]:
# Periksa koordinat ekstrem
extreme_coords = geolocations_df[
    (geolocations_df['geolocation_lng'] < -75) | 
    (geolocations_df['geolocation_lng'] > -30) |
    (geolocations_df['geolocation_lat'] < -35) | 
    (geolocations_df['geolocation_lat'] > 5)
]
print(f"Jumlah koordinat di luar rentang Brasil: {extreme_coords.shape[0]}")
print("Sampel 5 koordinat ekstrem:")
print(extreme_coords.sample(5))

Jumlah koordinat di luar rentang Brasil: 29
Sampel 5 koordinat ekstrem:
        geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
992584                        98780       -36.605374       -64.283946   
727755                        57319        45.065933         9.341528   
860562                        83252        42.184003        -8.723762   
516682                        28595        43.684961        -7.411080   
965687                        95130        14.585073       121.105394   

           geolocation_city geolocation_state  
992584           santa rosa                RS  
727755           pau d'arco                AL  
860562   ilha dos valadares                PR  
516682              portela                RJ  
965687  santa lucia do piai                RS  


**Insight:**
- Dari 1.000.163 baris data, terdapat 29 koordinat yang berada di luar rentang normal Brasil. Meskipun jumlahnya kecil (kurang dari 0,003% dari dataset), anomali ini perlu ditangani.
- Anomali yang terlihat pada sampel:

- Baris 727755: Koordinat (45.065933, 9.341528) untuk "Pau D'arco, AL" sebenarnya berada di Eropa Utara (sekitar Italia), bukan di Alagoas, Brasil.
- Baris 860562: Koordinat (42.184003, -8.723762) untuk "Ilha dos Valadares, PR" berada di sekitar Spanyol, bukan di Paraná, Brasil.
- Baris 516682: Koordinat (43.684961, -7.411080) untuk "Portela, RJ" juga berada di sekitar Spanyol, bukan di Rio de Janeiro.
- Baris 965687: Koordinat (14.585073, 121.105394) untuk "Santa Lucia do Piai, RS" berada di sekitar Filipina, sangat jauh dari Rio Grande do Sul.
- Baris 992584: Koordinat (-36.605374, -64.283946) untuk "Santa Rosa, RS" berada di Patagonia Argentina, bukan di Rio Grande do Sul, Brasil.
- Semua lokasi ini seharusnya berada di Brasil berdasarkan kode negara bagian (RS, AL, PR, RJ) yang merupakan kode resmi negara bagian Brasil, namun koordinatnya secara geografis berada di luar Brasil.
- pertimbangkan untuk menghapus 29 baris anomali ini karena jumlahnya sangat kecil dibandingkan total dataset.

In [47]:
customers_df.info()
customers_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


customer_zip_code_prefix
count              99441.000000
mean               35137.474583
std                29797.938996
min                 1003.000000
25%                11347.000000
50%                24416.000000
75%                58900.000000
max                99990.000000

In [48]:
# Periksa distribusi panjang digit kode pos pelanggan
digit_counts = customers_df['customer_zip_code_prefix'].astype(str).str.len().value_counts()
print("Distribusi panjang digit kode pos pelanggan:")
print(digit_counts)

# Periksa perbandingan jumlah customer_id vs customer_unique_id
unique_customer_ids = customers_df['customer_id'].nunique()
unique_customer_unique_ids = customers_df['customer_unique_id'].nunique()
print(f"Jumlah customer_id unik: {unique_customer_ids}")
print(f"Jumlah customer_unique_id unik: {unique_customer_unique_ids}")
print(f"Rasio: {unique_customer_ids / unique_customer_unique_ids:.2f}")

# Periksa distribusi pelanggan per negara bagian
state_counts = customers_df['customer_state'].value_counts()
print("\nTop 10 negara bagian berdasarkan jumlah pelanggan:")
print(state_counts.head(10))

# Periksa apakah ada kode pos dengan konsentrasi pelanggan yang sangat tinggi
zipcode_counts = customers_df['customer_zip_code_prefix'].value_counts()
print("\nTop 10 kode pos berdasarkan jumlah pelanggan:")
print(zipcode_counts.head(10))

Distribusi panjang digit kode pos pelanggan:
customer_zip_code_prefix
5    75446
4    23995
Name: count, dtype: int64
Jumlah customer_id unik: 99441
Jumlah customer_unique_id unik: 96096
Rasio: 1.03

Top 10 negara bagian berdasarkan jumlah pelanggan:
customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
Name: count, dtype: int64

Top 10 kode pos berdasarkan jumlah pelanggan:
customer_zip_code_prefix
22790    142
24220    124
22793    121
24230    117
22775    110
29101    101
13212     95
35162     93
22631     89
38400     87
Name: count, dtype: int64


In [51]:
# Periksa customer_unique_id yang memiliki lebih dari satu customer_id
duplicate_customers = customers_df.groupby('customer_unique_id')['customer_id'].count().reset_index()
customers_with_multiple_ids = duplicate_customers[duplicate_customers['customer_id'] > 1]
print(f"Jumlah pelanggan dengan multiple customer_id: {len(customers_with_multiple_ids)}")
print(f"Maksimum customer_id yang dimiliki satu pelanggan: {customers_with_multiple_ids['customer_id'].max()}")

# Sampel beberapa pelanggan dengan multiple IDs
print("\nSampel pelanggan dengan multiple customer_id:")
sample_multiple_ids = customers_with_multiple_ids.sample(5)['customer_unique_id'].tolist()
for cust_id in sample_multiple_ids:
    print(f"\nDetail untuk customer_unique_id: {cust_id}")
    print(customers_df[customers_df['customer_unique_id'] == cust_id][['customer_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']])

# Periksa konsistensi data geografis untuk customer_unique_id yang sama
customers_geo_consistency = customers_df.groupby('customer_unique_id').agg({
    'customer_zip_code_prefix': 'nunique',
    'customer_city': 'nunique',
    'customer_state': 'nunique'
}).reset_index()

inconsistent_geo = customers_geo_consistency[
    (customers_geo_consistency['customer_zip_code_prefix'] > 1) | 
    (customers_geo_consistency['customer_city'] > 1) | 
    (customers_geo_consistency['customer_state'] > 1)
]

print(f"\nJumlah pelanggan dengan data geografis tidak konsisten: {len(inconsistent_geo)}")
print(f"Persentase: {len(inconsistent_geo) / len(customers_geo_consistency) * 100:.2f}%")

Jumlah pelanggan dengan multiple customer_id: 2997
Maksimum customer_id yang dimiliki satu pelanggan: 17

Sampel pelanggan dengan multiple customer_id:

Detail untuk customer_unique_id: a00a13ac36037d575bb016caf8f50d53
                            customer_id  customer_zip_code_prefix  \
4378   f9accb2d97dc33b75ccc658df7e89236                     29072   
98633  e9285513ca554becdce6eb1b34963643                     29072   

      customer_city customer_state  
4378        vitoria             ES  
98633       vitoria             ES  

Detail untuk customer_unique_id: 33b239671c1c878914ed922a0dacd607
                            customer_id  customer_zip_code_prefix  \
80838  1ad00acd47366f0fad9b5258e2ef8a36                      5128   
81528  f26a3ca1beb5430f03e91621c7428b46                      5128   

      customer_city customer_state  
80838     sao paulo             SP  
81528     sao paulo             SP  

Detail untuk customer_unique_id: cde37fe3a24ad875f5ff1482fd80a43e
         

**Insight:**
- Data Relatif Bersih: Persentase pelanggan dengan data geografis tidak konsisten sangat kecil (0,26%)
- Multiple IDs dengan Alamat Sama: Ini menunjukkan bahwa pelanggan mungkin membuat akun baru untuk alasan lain (misalnya lupa password, ingin mendapatkan promo untuk pengguna baru, dll.)

In [56]:
product_categories.info()
product_categories.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


product_category_name product_category_name_english
count                     71                            71
unique                    71                            71
top             beleza_saude                 health_beauty
freq                       1                             1

**Insight:**
- Adanya 71 kategori produk berbeda menunjukkan platform e-commerce ini memiliki diversifikasi produk yang luas, mencakup berbagai segmen pasar.
- Dapat digunakan untuk segmentasi produk dan analisis performa lintas kategori.

In [58]:
products_df.info()
products_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


product_name_lenght  product_description_lenght  product_photos_qty  \
count         32341.000000                32341.000000        32341.000000   
mean             48.476949                  771.495285            2.188986   
std              10.245741                  635.115225            1.736766   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32949.000000       32949.000000       32949.000000   
mean        2276.472488          30.815078          16.937661   
std         4282.038731          16.914458          13.637554   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1900.000000          38.000000          21.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count      32949.000000  
mean          23.196728  
std           12.079047  
min            6.000000  
25%           15.000000  
50%           20.000000  
75%           30.000000  
max          118.000000

**Insight:**
- Missing Values:product_category_name: 610 nilai kosong (32951 - 32341 = 610)
product_name_lenght, product_description_lenght, dan product_photos_qty: 610 nilai kosong
product_weight_g, product_length_cm, product_height_cm, dan product_width_cm: 2 nilai kosong (32951 - 32949 = 2)
- Kolom product_name_lenght dan product_description_lenght memiliki kesalahan ejaan ("lenght" bukan "length")
- Deskripsi produk dengan hanya 4 karakter sangat pendek dan tidak memadai untuk menggambarkan produk dengan baik
- Banyak produk hanya memiliki 1 foto, yang kurang optimal untuk e-commerce
- Nilai minimum 0g tidak masuk akal untuk produk fisik
- Nilai maksimum 40425g (40.4kg) sangat berat, produk khusus atau kesalahan input
- Beberapa produk memiliki dimensi sangat besar (maksimum 105cm untuk tinggi dan panjang, 118cm untuk lebar)
- Semua atribut dimensi dan kuantitas menggunakan tipe data float64, yang mungkin tidak efisien untuk data yang seharusnya berupa integer (seperti jumlah foto)


**Insight:**
- xxx
- xxx

### Cleaning Data

**sellers_df**

In [49]:
# Normalisasi kode pos menjadi format 5 digit dengan padding 0
sellers_df['seller_zip_code_prefix_normalized'] = sellers_df['seller_zip_code_prefix'].astype(str).str.zfill(5)

In [50]:
# Cek apakah kode pos 4 digit terdapat dalam dataset geolokasi
four_digit_zips = sellers_df[sellers_df['seller_zip_code_prefix'].astype(str).str.len() == 4]['seller_zip_code_prefix'].unique()
matching_geo = geolocations_df[geolocations_df['geolocation_zip_code_prefix'].isin(four_digit_zips)]
print(f"Jumlah kode pos 4 digit yang ditemukan dalam dataset geolokasi: {matching_geo['geolocation_zip_code_prefix'].nunique()}")

Jumlah kode pos 4 digit yang ditemukan dalam dataset geolokasi: 829


**geolocations_df**

In [ ]:
# Identifikasi semua koordinat geolocation yang tidak valid
invalid_coords = geolocations_df[
    (geolocations_df['geolocation_lng'] < -75) | 
    (geolocations_df['geolocation_lng'] > -30) |
    (geolocations_df['geolocation_lat'] < -35) | 
    (geolocations_df['geolocation_lat'] > 5)
]

# Simpan daftar koordinat tidak valid untuk referensi
invalid_coords.to_csv('data/invalid_coordinates.csv', index=False)

# Buat versi dataset yang sudah dibersihkan
geolocations_clean_df = geolocations_df[~geolocations_df.index.isin(invalid_coords.index)]

# Verifikasi ukuran dataset baru
print(f"Ukuran dataset original: {geolocations_df.shape[0]} baris")
print(f"Ukuran dataset dibersihkan: {geolocations_clean_df.shape[0]} baris")
print(f"Jumlah baris yang dihapus: {geolocations_df.shape[0] - geolocations_clean_df.shape[0]} baris")

Ukuran dataset original: 1000163 baris
Ukuran dataset dibersihkan: 1000134 baris
Jumlah baris yang dihapus: 29 baris


**customers_df**

In [52]:
# Standarisasi semua kode pos menjadi format 5 digit
customers_df['customer_zip_code_prefix_norm'] = customers_df['customer_zip_code_prefix'].astype(str).str.zfill(5)

In [53]:
# Buat flag untuk pelanggan dengan multiple IDs
customer_id_counts = customers_df.groupby('customer_unique_id')['customer_id'].count().reset_index()
customer_id_counts.rename(columns={'customer_id': 'id_count'}, inplace=True)
customers_df = customers_df.merge(customer_id_counts, on='customer_unique_id', how='left')
customers_df['has_multiple_ids'] = customers_df['id_count'] > 1

**products_df**


In [60]:
#penanganan missing values
# Cek proporsi missing values untuk setiap kolom
missing_values = products_df.isnull().sum()
missing_percentage = (missing_values / len(products_df)) * 100
print(missing_percentage)

# Analisis pola missing values
print("Jumlah produk tanpa kategori:", products_df['product_category_name'].isnull().sum())

# Periksa apakah missing values pada beberapa kolom saling terkait
missing_pattern = products_df[products_df['product_category_name'].isnull()]
print("Jumlah nilai kosong pada atribut lain untuk produk tanpa kategori:")
print(missing_pattern.isnull().sum())

product_id                    0.000000
product_category_name         1.851234
product_name_lenght           1.851234
product_description_lenght    1.851234
product_photos_qty            1.851234
product_weight_g              0.006070
product_length_cm             0.006070
product_height_cm             0.006070
product_width_cm              0.006070
dtype: float64
Jumlah produk tanpa kategori: 610
Jumlah nilai kosong pada atribut lain untuk produk tanpa kategori:
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                1
product_length_cm               1
product_height_cm               1
product_width_cm                1
dtype: int64


In [61]:
# Perbaiki nama kolom dengan ejaan yang salah
products_df = products_df.rename(columns={
    'product_name_lenght': 'product_name_length',
    'product_description_lenght': 'product_description_length'
})

In [65]:
# penanganan outlier berat
# Identifikasi produk dengan berat 0g
zero_weight = products_df[products_df['product_weight_g'] == 0]
print(f"Jumlah produk dengan berat 0g: {len(zero_weight)}")

# Identifikasi produk dengan berat sangat besar
heavy_products = products_df[products_df['product_weight_g'] > 10000]
print(f"Jumlah produk dengan berat >10kg: {len(heavy_products)}")
print(heavy_products[['product_id', 'product_category_name', 'product_weight_g']].head())

Jumlah produk dengan berat 0g: 4
Jumlah produk dengan berat >10kg: 1891
                          product_id      product_category_name  \
6   732bd381ad09e530fe0a5f457d81becb                 cool_stuff   
44  7a8dac4aaa16bc642e4df33adcf03303                 cool_stuff   
46  5f504b3a1c75b73d6151be81eb05bdc9                 cool_stuff   
68  1e5428c428e0f783acd6e3d94ba4ee2a                        pcs   
79  07f01b6fcacc1b187a71e5074199db2d  agro_industria_e_comercio   

    product_weight_g  
6            18350.0  
44           11750.0  
46           15350.0  
68           12000.0  
79           16400.0  


In [67]:
# Identifikasi produk dengan deskripsi sangat pendek
short_desc = products_df[products_df['product_description_length'] < 20]
print(f"Jumlah produk dengan deskripsi <20 karakter: {len(short_desc)}")

Jumlah produk dengan deskripsi <20 karakter: 7


In [68]:
# Hitung volume produk sebagai metrik tambahan
products_df['product_volume_cm3'] = products_df['product_length_cm'] * products_df['product_height_cm'] * products_df['product_width_cm']

# Identifikasi produk dengan volume ekstrem
large_volume = products_df[products_df['product_volume_cm3'] > 100000]
print(f"Jumlah produk dengan volume >100.000 cm³: {len(large_volume)}")

Jumlah produk dengan volume >100.000 cm³: 715


In [69]:
# Analisis statistik berdasarkan kategori
category_stats = products_df.groupby('product_category_name').agg({
    'product_weight_g': ['mean', 'median', 'std', 'min', 'max'],
    'product_volume_cm3': ['mean', 'median', 'std'],
    'product_photos_qty': ['mean', 'median']
})

# Identifikasi kategori dengan karakteristik ekstrem
print(category_stats.sort_values(('product_weight_g', 'mean'), ascending=False).head())

                                               product_weight_g           \
                                                           mean   median   
product_category_name                                                      
moveis_colchao_e_estofado                          13190.000000   8675.0   
moveis_escritorio                                  12740.867314  11600.0   
moveis_cozinha_area_de_servico_jantar_e_jardim     11598.563830   9950.0   
moveis_quarto                                       9997.222222  10600.0   
eletrodomesticos_2                                  9913.333333   9050.0   

                                                                               \
                                                         std     min      max   
product_category_name                                                           
moveis_colchao_e_estofado                       12336.391513  1700.0  30000.0   
moveis_escritorio                                6106.700414    50.

## Exploratory Data Analysis (EDA)

### Explore ...

**Insight:**
- xxx
- xxx

## Visualization & Explanatory Analysis

### Pertanyaan 1:

### Pertanyaan 2:

**Insight:**
- xxx
- xxx

## Analisis Lanjutan (Opsional)

## Conclusion

- Conclution pertanyaan 1
- Conclution pertanyaan 2